src/app/clusters/tls-client-management-server/tls-client-management-server.cpp

```diff
@@ -0,0 +1,244 @@
+{
+    for (uint8_t i = 0; true; i++)
+    {
+        TLSEndpointStruct::Type endpoint;
+
+        auto err = mDelegate.GetProvisionedEndpointByIndex(i, endpoint);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(err);
+
+        ReturnErrorOnFailure(encoder.Encode(endpoint));
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TlsClientManagementServer::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    VerifyOrDie(aPath.mClusterId == TlsClientManagement::Id);
+    Status status;
+
+    switch (aPath.mAttributeId)
+    {
+    case MaxProvisioned::Id: {
+        uint8_t newValue;
+        ReturnErrorOnFailure(aDecoder.Decode(newValue));
+        status = SetMaxProvisioned(newValue);
+        return StatusIB(status).ToChipError();
+    }
+    default:
+        // Unknown attribute
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+}
+
+Status TlsClientManagementServer::SetMaxProvisioned(uint8_t maxProvisioned)
+{
+    bool enableChanged = mMaxProvisioned != maxProvisioned;
+
+    if (enableChanged)
+    {
+        mMaxProvisioned = maxProvisioned;
+
+        // Write new value to persistent storage.
+        auto endpointId            = GetEndpointId();
+        ConcreteAttributePath path = ConcreteAttributePath(endpointId, TlsClientManagement::Id, MaxProvisioned::Id);
+        GetSafeAttributePersistenceProvider()->WriteScalarValue(path, mMaxProvisioned);
+
+        // and mark as dirty
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return Protocols::InteractionModel::Status::Success;
+}
+
+void TlsClientManagementServer::InvokeCommand(HandlerContext & ctx)
+{
+    switch (ctx.mRequestPath.mCommandId)
+    {
+    case Commands::ProvisionEndpoint::Id:
+        CommandHandlerInterface::HandleCommand<Commands::ProvisionEndpoint::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleProvisionEndpoint(innerCtx, req); });
+        break;
+    case Commands::FindEndpoint::Id:
+        CommandHandlerInterface::HandleCommand<Commands::FindEndpoint::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleFindEndpoint(innerCtx, req); });
+        break;
+    case Commands::RemoveEndpoint::Id:
+        CommandHandlerInterface::HandleCommand<Commands::RemoveEndpoint::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleRemoveEndpoint(innerCtx, req); });
+        break;
+    }
+}
+
+
+void TlsClientManagementServer::HandleProvisionEndpoint(HandlerContext & ctx, const Commands::ProvisionEndpoint::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsClientManagement: ProvisionEndpoint");
+
+    Commands::ProvisionEndpointResponse::Type response;
+    Status status = mDelegate.ProvisionEndpoint(req, response.endpointID);
```

Comment:
This should do constraint enforcement (Hostname length limits, CAID not allowed to be too large) before passing the data to the delegate.